dali/imgcodec/parsers/webp.cc

```diff
@@ -12,19 +13,158 @@
+  return header.riff_text == tag("RIFF") && header.webp_text == tag("WEBP");
+}
+
+bool is_simple_lossy_format(const vp8_header_t &header) {
+  return header == tag("VP8 ");
+}
+
+bool is_simple_lossless_format(const vp8_header_t &header) {
+  return header == tag("VP8L");
+}
+
+bool is_extended_format(const vp8_header_t &header) {
+  return header == tag("VP8X");
+}
+
+bool is_sync_code_valid(const WebpLossyHeader &header) {
+  return header.sync_code == std::array<uint8_t, 3>{0x9D, 0x01, 0x2A};
+}
+
+bool is_sync_code_valid(const WebpLosslessHeader &header) {
+  return header.signature_byte == 0x2F;
+}
+
+template<size_t N>
+std::string sequence_of_integers(const std::array<uint8_t, N> &data) {
+  std::string result;
+  for (size_t i = 0; i < N; i++)
+    result += (i == 0 ? "" : " ") + std::to_string(data[i]);
+  return result;
+}
+
+}  // namespace
+
 ImageInfo WebpParser::GetInfo(ImageSource *encoded) const {
+  auto stream = encoded->Open();
   ImageInfo info;
-  // info.shape = {h, w, c};   // TODO(janton)
+
+  stream->Skip<RiffHeader>();
+  const auto vp8_header = stream->ReadOne<vp8_header_t>();
+  if (is_simple_lossy_format(vp8_header)) {
+    const auto lossy_header = stream->ReadOne<WebpLossyHeader>();
+    if (!is_sync_code_valid(lossy_header)) {
+      DALI_FAIL("Sync code 157 1 42 not found at expected position. Found " +
+                sequence_of_integers(lossy_header.sync_code) + " instead");
+    }
+
+    const int w = ReadValueLE<uint16_t>(*stream) & 0x3FFF;
+    const int h = ReadValueLE<uint16_t>(*stream) & 0x3FFF;
+
+    // VP8 always uses RGB
+    info.shape = {h, w, 3};
+  } else if (is_simple_lossless_format(vp8_header)) {
+    const auto lossless_header = stream->ReadOne<WebpLosslessHeader>();
+    if (!is_sync_code_valid(lossless_header)) {
+        DALI_FAIL("Sync code 47 not found at expected position. Found " +
+                  std::to_string(lossless_header.signature_byte) + " instead.");
+    }
+
+    // VP8L shape information starts after the sync code
+    const auto features = ReadValueLE<uint32_t>(*stream);
+    const int w = (features & 0x00003FFF) + 1;
+    const int h = ((features & 0x0FFFC000) >> 14) + 1;
+    const int alpha = (features & 0x10000000) >> 28;
+
+    // VP8L always uses RGBA
+    info.shape = {h, w, 3 + alpha};
+  } else if (is_extended_format(vp8_header)) {
+    DALI_FAIL("WebP extended file format is not supported.");
+  } else {
+    DALI_FAIL("Unrecognized WebP header: " + sequence_of_integers(vp8_header));
+  }
+
   return info;
 }
 
 bool WebpParser::CanParse(ImageSource *encoded) const {
-  return false;  // TODO(janton)
+  static_assert(sizeof(WebpLossyHeader) > sizeof(WebpLosslessHeader));
+  uint8_t data[sizeof(RiffHeader) + sizeof(vp8_header_t) + sizeof(WebpLossyHeader)];
+  if (!ReadHeader(data, encoded, sizeof(data)))
+    return false;
+  MemInputStream stream(data, sizeof(data));
+
+  if (!is_valid_riff_header(stream.ReadOne<RiffHeader>()))
```

Comment:
It's not perfectly kosher, but since we're checking only the tag, it should be OK without endianness fixup.